Lib/unittest/mock.py

```diff
@@ -2054,6 +2054,10 @@ def __eq__(self, other):
         else:
             self_name, self_args, self_kwargs = self
 
+        if (getattr(self, 'parent', None) and getattr(other, 'parent', None)
```

Comment:
`self` should always have a parent attribute, no need to use getattr, isn't it?

You are using gettattr to ignore the value and then compare it. What if it has a parent but it evaluates Falsy? Might be better to just  use `hasattr`.

```python
if hasattr(other, 'parent') and self.parent != other.parent:
    return False
```

I think it reads easier

Reply:
The implementation was suggested in the bug discussion, @voidspace can make a call here.

Reply:
I think @mariocj89 has a point. On `call()` the parent is `None`.

Reply:
Is `hasattr` safe to use here or does it still risk swallowing exceptions other than `AttributeError` on Python 3.6+? The grumpy old python dev in me still prefers `getattr(other, 'parent', None)` out of paranoia...

Reply:
`hasattr` was fixed quite some time ago I believe :-)

Reply:
Interestingly, bunch of tests fail after the changes suggested here...

Reply:
@mariocj89 / @voidspace - I'm afraid with the suggested changes, we get a bunch of failure:
https://travis-ci.org/python/cpython/jobs/455556002#L2849-L3008
I don't *think* I've done anything wrong here. I played around locally, and with anything other exactly as it was originally, you get at least one test failure...

Reply:
Wow, that's unexpected :-)

Reply:
Indeed! @voidspace - are you happy to go with the original solution so we can get this merged?

Reply:
Surprising indeed!

Look at this:
```python
>>> from unittest.mock import Mock
>>> m = Mock()
>>> m.abc.cde()
<Mock name='mock.abc.cde()' id='139944355306408'>
>>> m.mock_calls
[call.abc.cde()]
>>> print(m.mock_calls[0].parent)
None
```

Seems parent is not set at all, which is causing a lot of tests to fail with my proposal :/

Reply:
Wow. That seems like a bug too, let me add some cases and see about  a fix.
@voidspace - open question: shall we keep `getattr(other, 'parent', None)` to cover the porting-to-testing-cabal version issue?

Reply:
(just as a note, we can follow this up indeed in another issue as I am not sure about the fix, seems we are not passing the parent here): https://github.com/python/cpython/blob/master/Lib/unittest/mock.py#L986)

Reply:
I'm afraid it's intrinsically linked to this issue, `call` children are designed to be compared to items in `Mock().method_calls` and `Mock().mock_call` and I don't think that currently works safely... It's generally always be true, which isn't what we want with nested/generative calls...

Reply:
Where there's a single call there's no parent call. So having parent not set in that example looks correct.